MicroPython/src/cloud/KAA_cloud.py

```diff
@@ -0,0 +1,190 @@
+            return 1
+
+        config.cfg.ap_config_done = True
+        config.cfg.save()
+        machine.reset()
+
+        return 0
+
+    def configure_data(self) -> None:
+        """
+        Setup data from kaa_config file and save it to general config file
+        :return: None
+        """
+        logging.debug("KAA_cloud/configure_data()")
+        kaa_configuration = self.load_kaa_config_from_file()
+        
+        config.cfg.kaa_user = kaa_configuration.get(
+            "kaa_user", config.DEFAULT_KAA_USER)
+
+        config.cfg.kaa_password = kaa_configuration.get(
+            "kaa_password", config.DEFAULT_KAA_PASSWORD)
+
+        config.cfg.kaa_app_version = kaa_configuration.get(
+            "kaa_app_version", config.DEFAULT_KAA_APP_VERSION)
+        
+        config.cfg.kaa_endpoint = kaa_configuration.get(
+            "kaa_endpoint", config.DEFAULT_KAA_APP_VERSION)
+
+        # Update in case app_version of kaa_endpoint changed
+        config.cfg.kaa_topic = 'kp1/{}/dcx/{}/json/{}'.format(
+            config.cfg.kaa_app_version, 
+            config.cfg.kaa_endpoint, 
+            config.cfg.mqqt_request_id
+        )
+
+        self.publish_success_topic = config.cfg.kaa_topic + '/status'
+        self.publish_error_topic = config.cfg.kaa_topic + '/error'
+
+        config.cfg.save()
+
+    def load_kaa_config_from_file(self) -> dict:
+        """
+        Load configuration of AWS from file.
+        :return: Configuration in dict.
+        """
+        if utils.check_if_file_exists(config.KAA_CONFIG_PATH) == 0:
+            raise Exception("Create kaa_config.json file first!")
+
+        with open(config.KAA_CONFIG_PATH, "r", encoding="utf8") as infile:
+            config_dict = ujson.load(infile)
+
+        return config_dict
+
+    def _format_data(self, data: dict) -> dict:
+        """
+        Helper function for formatting data to match Kaa expected input
+        :param data: Data in dict to be formatted
+        :return dict: Formatted data
+        """
+        formatted_data = {}
+        for ind, (key, values) in enumerate(data.items()):
+            # Unpack outer list and extract values to variables
+            (_, value), = values
+            formatted_data[key] = value
+        
+        return formatted_data
+
+    def publish_data(self, data):
+        wireless_controller, mqtt_communicator = utils.get_wifi_and_cloud_handlers(
+            sync_time=False
+        )
+
+        result_suc_topic = mqtt_communicator.subscribe(
+            topic=self.publish_success_topic, callback=self.receive_message, qos=config.cfg.QOS
+        )
+        result_err_topic = mqtt_communicator.subscribe(
+            topic=self.publish_error_topic, callback=self.receive_message, qos=config.cfg.QOS
+        )
+        if not result_suc_topic or not result_err_topic:
+            logging.error(
+                "Error subscribing to topics with MQTT in publish_data()")
+
+        # TODO: Certificates?
```

Comment:
?

Reply:
We're currently unable to create secure connection through SSL as Kaa does not allow us to create certificates (internal server error during generating of a key) so there are no certificates for now.